hypervisor/openstack: fix usage with V3 API

GiedriusS opened the pull request:
KeyStone V3 API needs the project name instead of the tenant's name
(such field doesn't exist anymore). Add proper parameter checks:
* With `v3`, `openstack_project_name` is expected
* With `v2`, `openstack_tenant` is expected (like before)

Also, in our environment no floating IP pools are available so I've also added an option which controls whether a floating IP should be allocated for the machines on which tests are ran.

Fix one issue with changed naming + bump `fog-openstack` requirement to `~> 1.0.0`.

Verification: spec and acceptance tests pass.

Commit:
hypervisor/openstack: permit specifying openstack_project_name

KeyStone V3 API needs the project name instead of the tenant's name
(such field doesn't exist anymore). Transparently let users pass both
options. At least one needs to be specified. Later on, Fog handles
switching between the V3 and V2 versions depending on the
openstack_auth_url parameter.

Commit:
hypervisor/openstack: make it work with newest fog-openstack

The errors have been renamed to Fog::OpenStack::* instead of
Fog::Compute::OpenStack::*. At the same time bump the requirement of
`fog-openstack` because it will not work with earlier versions.

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

Commit:
Revert "hypervisor/openstack: make it work with newest fog-openstack"

This reverts commit e75b94caa8255b7802cbb3d330255801e9a0ff5f.

Commit:
beaker-openstack.gemspec: pin version which works

kevpl commented on lib/beaker/hypervisor/openstack.rb:
```diff
@@ -43,7 +49,7 @@ def initialize(openstack_hosts, options)
         :openstack_tenant   => @options[:openstack_tenant],
         :openstack_region   => @options[:openstack_region],
-      }
+    }.merge(extra_credentials)
```
this line should be intended once more to match the block it's surrounding.

kevpl reviewed (changes requested):
one small change requested in current code. Please add some spec tests to ensure we don't break your fix in the future, & we'll look to push this through testing to release!

GiedriusS commented:
Thanks for the quick review! Will definitely do that and ping you once the v2 is ready.

Commit:
beaker: add proper check for args

With v2 only certain parameters are allowed, and with v3 - different
ones. Add checks for those. Refactor the v3 check into a different
variable. Fix the tests so that they would pass with the new checks.

Commit:
spec/beaker: add missing test for nil

This field should be nil too so add a test for it.

Commit:
beaker/openstack: print out more parameters on err

Commit:
beaker/openstack: make creating floating IPs optional

* Move `get_ip` to `get_floating_ip`
* Add another parameter `openstack_floating_ip` (bool) which controls if
we want to create a floating IP address for the VM
* If no floating IP is requested, generate a random 10 character
hostname

kevpl reviewed (approved):
apologies on taking so long to come back to this, lgtm 👍. Let's get this through testing:

@puppetlabs-jenkins test this please

@GiedriusS, from your comment at the top, it looks like this change is backwards compatible, in that you can still use the previous version of the API (v2) without changing any of your beaker-openstack settings. Is this correct, or am I not understanding your comment at the beginning of the PR on usage? I'm looking to figure out if this change is backwards-compatible or not for our version number update on release.

GiedriusS commented:
From OpenStack's perspective it is not but from our perspective - it is because we can still pass the V2 parameters to `fog-openstack` and then later on it should use only V2 API calls. You can find more here:
https://docs.openstack.org/keystone/pike/contributor/http-api.html

> 
> Note a few key differences when compared to the v2.0 API:
> 
> A “tenant” in v2.0 became a “project” in v3.

`fog-openstack` supports both: https://github.com/fog/fog-openstack/tree/master/lib/fog/openstack/identity

GiedriusS commented:
FWIW, I have only tested the happy paths that were added in this PR. Have been using it for about a month without any problems. Can't test it with `v2` since I don't have any endpoints to play around with :smile: 

AFAICT the only non-backward compatible change is the requirement to pass the parameter `openstack_floating_ip` so there should be a new major version.

kevpl commented:
test this please

kevpl commented:
released as beaker-openstack [0.3.0](https://github.com/puppetlabs/beaker-openstack/compare/0.2.0...0.3.0)! Thanks for your contribution @GiedriusS!

GiedriusS commented:
Thanks, @kevpl, switched back to using the new upstream version here internally!
